File: backend/internal/models/material.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Material struct {
	ID              uuid.UUID `json:"id" db:"id"`
	TemaID          uuid.UUID `json:"tema_id" db:"tema_id"`
	Titulo          string    `json:"titulo" db:"titulo"`
	Tipo            string    `json:"tipo" db:"tipo"` // pdf, video, link, documento, imagen
	URLArchivo      string    `json:"url_archivo" db:"url_archivo"`
	TamanoMB        *float64  `json:"tamano_mb,omitempty" db:"tamano_mb"`
	DuracionMinutos *int      `json:"duracion_minutos,omitempty" db:"duracion_minutos"`
	Descripcion     *string   `json:"descripcion,omitempty" db:"descripcion"`
	Orden           int       `json:"orden" db:"orden"`
	Activo          bool      `json:"activo" db:"activo"`
	CreatedAt       time.Time `json:"created_at" db:"created_at"`

	// Para estudiante
	VistoPorMi bool `json:"visto_por_mi,omitempty"`

	// Para docente
	CantidadVistos   int `json:"cantidad_vistos,omitempty"`
	TotalEstudiantes int `json:"total_estudiantes,omitempty"`
}

type CreateMaterialRequest struct {
	TemaID          uuid.UUID `json:"tema_id" binding:"required"`
	Titulo          string    `json:"titulo" binding:"required"`
	Tipo            string    `json:"tipo" binding:"required,oneof=pdf video link documento imagen"`
	URLArchivo      string    `json:"url_archivo" binding:"required"`
	TamanoMB        *float64  `json:"tamano_mb"`
	DuracionMinutos *int      `json:"duracion_minutos"`
	Descripcion     *string   `json:"descripcion"`
	Orden           int       `json:"orden" binding:"required,min=1"`
}

type MaterialVisto struct {
	ID           uuid.UUID `json:"id" db:"id"`
	MaterialID   uuid.UUID `json:"material_id" db:"material_id"`
	EstudianteID uuid.UUID `json:"estudiante_id" db:"estudiante_id"`
	VistoEn      time.Time `json:"visto_en" db:"visto_en"`
}

// ✅ AGREGAR ESTE STRUCT A backend/internal/models/material.go

type UpdateMaterialRequest struct {
	Titulo      *string  `json:"titulo,omitempty"`
	Tipo        *string  `json:"tipo,omitempty"`
	URLArchivo  *string  `json:"url_archivo,omitempty"`
	Descripcion *string  `json:"descripcion,omitempty"`
	TamanoMB    *float64 `json:"tamano_mb,omitempty"`
	Orden       *int     `json:"orden,omitempty"`
}

File: frontend/recetario/lib/data/models/material.dart
class Material {
  final String id;
  final String temaId;
  final String titulo;
  final String tipo; // pdf, video, link, documento, imagen
  final String urlArchivo;
  final double? tamanoMb;
  final int? duracionMinutos;
  final String? descripcion;
  final int orden;
  final bool activo;
  final DateTime createdAt;
  
  // Para estudiante
  final bool? vistoPorMi;
  
  // Para docente
  final int? cantidadVistos;
  final int? totalEstudiantes;

  Material({
    required this.id,
    required this.temaId,
    required this.titulo,
    required this.tipo,
    required this.urlArchivo,
    this.tamanoMb,
    this.duracionMinutos,
    this.descripcion,
    required this.orden,
    required this.activo,
    required this.createdAt,
    this.vistoPorMi,
    this.cantidadVistos,
    this.totalEstudiantes,
  });

  factory Material.fromJson(Map<String, dynamic> json) {
    return Material(
      id: json['id'],
      temaId: json['tema_id'],
      titulo: json['titulo'],
      tipo: json['tipo'],
      urlArchivo: json['url_archivo'],
      tamanoMb: json['tamano_mb']?.toDouble(),
      duracionMinutos: json['duracion_minutos'],
      descripcion: json['descripcion'],
      orden: json['orden'],
      activo: json['activo'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
      vistoPorMi: json['visto_por_mi'],
      cantidadVistos: json['cantidad_vistos'],
      totalEstudiantes: json['total_estudiantes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tema_id': temaId,
      'titulo': titulo,
      'tipo': tipo,
      'url_archivo': urlArchivo,
      'tamano_mb': tamanoMb,
      'duracion_minutos': duracionMinutos,
      'descripcion': descripcion,
      'orden': orden,
    };
  }

  String get tipoIcono {
    switch (tipo) {
      case 'pdf':
        return '📄';
      case 'video':
        return '🎥';
      case 'link':
        return '🔗';
      case 'imagen':
        return '🖼️';
      default:
        return '📎';
    }
  }

  String get duracionFormateada {
    if (duracionMinutos == null) return '';
    if (duracionMinutos! < 60) return '${duracionMinutos}min';
    final horas = duracionMinutos! ~/ 60;
    final mins = duracionMinutos! % 60;
    return '${horas}h ${mins}min';
  }

  String get tamanoFormateado {
    if (tamanoMb == null) return '';
    if (tamanoMb! < 1) return '${(tamanoMb! * 1024).toStringAsFixed(0)} KB';
    return '${tamanoMb!.toStringAsFixed(1)} MB';
  }
}